crates/polars-core/src/chunked_array/ops/flarion_funcs.rs

```diff
@@ -66,3 +89,333 @@ pub fn flarion_substring(s: &str, from: i32, len: i32) -> String {
+            PlSmallStr::EMPTY,
+            strings.len(),
+            &DataType::String,
+        ));
+    }
+
+    let result: StringChunked = match (from_length, lens_length) {
+        // Commented out because it is already covered in previous check
+        // (0, _) | (_, 0) => None,
+        (1, 1) => {
+            let mut builder = StringChunkedBuilder::new(PlSmallStr::EMPTY, strings.len());
+            for s_opt in strings.into_iter() {
+                let value = match (s_opt, first_from, first_len) {
+                    (Some(s), Some(from), Some(len)) => Some(flarion_substring(s, from, len)),
+                    (Some(s), Some(from), None) => substring_with_null_length(s, from),
+                    _ => None,
+                };
+                builder.append_option(value);
+            }
+            builder.finish()
+        },
+        (1, _) => {
+            let mut builder = StringChunkedBuilder::new(PlSmallStr::EMPTY, strings.len());
+            for (s_opt, len_opt) in strings.into_iter().zip(lens_iter) {
+                let value = match (s_opt, first_from, len_opt) {
+                    (Some(s), Some(from), Some(len)) => Some(flarion_substring(s, from, len)),
+                    (Some(s), Some(from), None) => substring_with_null_length(s, from),
+                    _ => None,
+                };
+                builder.append_option(value);
+            }
+            builder.finish()
+        },
+        (_, 1) => {
+            let mut builder = StringChunkedBuilder::new("".into(), strings.len());
+            for (s_opt, from_opt) in strings.into_iter().zip(from_iter) {
+                let value = match (s_opt, from_opt, first_len) {
+                    (Some(s), Some(from), Some(len)) => Some(flarion_substring(s, from, len)),
+                    (Some(s), Some(from), None) => substring_with_null_length(s, from),
+                    _ => None,
+                };
+                builder.append_option(value);
+            }
+            builder.finish()
+        },
+        _ => {
+            let mut builder = StringChunkedBuilder::new("".into(), strings.len());
+            for (s_opt, (from_opt, len_opt)) in strings
+                .into_iter()
+                .zip(from_iter.into_iter().zip(lens_iter))
+            {
+                let value = match (s_opt, from_opt, len_opt) {
+                    (Some(s), Some(from), Some(len)) => Some(flarion_substring(s, from, len)),
+                    (Some(s), Some(from), None) => substring_with_null_length(s, from),
+                    _ => None,
+                };
+                builder.append_option(value);
+            }
+            builder.finish()
+        },
+    };
+
+    Ok(result.into_series())
+}
+
+// I'm hoping the compiler optimizes this out, but it is required for correct outputs
+#[inline]
+fn fill_empty_iter<'a, I: Iterator<Item = &'a str>>(in_iter: I) -> impl Iterator<Item = &'a str> {
+    let chars = in_iter.collect::<Vec<_>>();
```

Comment:
more readble?

```rust
let chars = in_iter.collect::<Vec<_>>();
if chars.is_empty() {
    // Can't use iter::once() because both we expect Vec's specific IntoIter
    return vec![""].into_iter();
}
chars.into_iter()
```